File: shared/plural/src/plural/env.py
from typing import Self
from os import environ

from pydantic import BaseModel
from bson import ObjectId

from .missing import MISSING


INSTANCE = str(ObjectId())


class Env(BaseModel):
    bot_token: str
    discord_url: str
    redis_url: str
    mongo_url: str
    domain: str
    max_avatar_size: int
    dev: bool
    cdn_upload_token: str
    admins: set[int]
    patreon_secret: str
    info_bot_token: str

    @classmethod
    def new(cls) -> Self:
        return cls.model_validate({
            'bot_token': environ.get('BOT_TOKEN', MISSING),
            'discord_url': environ.get('DISCORD_URL', MISSING),
            'redis_url': environ.get('REDIS_URL', MISSING),
            'mongo_url': environ.get('MONGO_URL', MISSING),
            'domain': environ.get('DOMAIN', MISSING),
            'max_avatar_size': int(environ.get('MAX_AVATAR_SIZE', MISSING)),
            'dev': environ.get('DEV', '1') != '0',
            'cdn_upload_token': environ.get('CDN_UPLOAD_TOKEN', MISSING),
            'admins': (
                set(map(int, environ.get('ADMINS', '').split(',')))
                if environ.get('ADMINS') else set()),
            'patreon_secret': environ.get('PATREON_SECRET', ''),
            'info_bot_token': environ.get('INFO_BOT_TOKEN', '')
        })

    @property
    def avatar_url(self) -> str:
        return (
            f'https://cdn{'dev' if self.dev else ''}'
            f'.{self.domain}/images/{{parent_id}}/{{hash}}.webp'
        )


env = Env.new()

File: core/src/env.rs
use std::{env::var, sync::OnceLock};


static ENV: OnceLock<Env> = OnceLock::new();

#[derive(Debug)]
pub struct Env {
    pub bot_token: String,
    pub redis_url: String,
    pub mongo_url: String,
    pub domain: String,
    pub max_avatar_size: u32,
    pub dev: bool,
    pub internal_master_token: String,
    pub admins: Vec<u64>,
    pub patreon_secret: String,
    pub info_bot_token: String,
    // ? properties
    pub avatar_url: String
}

impl Default for Env {
    fn default() -> Self {
        let dev = cfg!(debug_assertions);

        let domain = var("DOMAIN").unwrap_or("example.com".to_string());

        Self {
            bot_token: var("BOT_TOKEN").expect("BOT_TOKEN is not set"),
            redis_url: var("REDIS_URL").expect("REDIS_URL is not set"),
            mongo_url: var("MONGO_URL").expect("MONGO_URL is not set"),
            domain: domain.clone(),
            max_avatar_size: var("MAX_AVATAR_SIZE")
                .unwrap_or("4194304".to_string())
                .parse()
                .expect("MAX_AVATAR_SIZE is not a valid number"),
            dev,
            internal_master_token: var("INTERNAL_MASTER_TOKEN")
                .expect("INTERNAL_MASTER_TOKEN is not set"),
            admins: var("ADMINS")
                .unwrap_or("".to_string())
                .split(',')
                .filter_map(|s| s.parse::<u64>().ok())
                .collect(),
            patreon_secret: var("PATREON_SECRET").unwrap_or("".to_string()),
            info_bot_token: var("INFO_BOT_TOKEN").unwrap_or("".to_string()),
            // ? properties
            avatar_url: format!(
                "https://cdn{}.{}/images/{{parent_id}}/{{hash}}.webp",
                if dev { "dev" } else { "" },
                domain
            )
        }
    }
}

impl Env {
    pub fn new() -> Self {
        Self::default()
    }
}

pub fn env() -> &'static Env {
    ENV.get_or_init(Env::new)
}
